File: algorithms/algorithms-cplusplus/leetcode/Question_0383_Ransom_Note.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  bool canConstruct(string ransomNote, string magazine) {
    if (ransomNote.size() > magazine.size()) {
      return false;
    }
    vector<int> index(256, 0);
    for (char c : magazine) {
      index[c]++;
    }
    for (char c : ransomNote) {
      if (index[c]-- <= 0) {
        return false;
      }
    }
    return true;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0383_Ransom_Note.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0383_Ransom_Note {
    public boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote == null || magazine == null || ransomNote.length() > magazine.length()) {
            return false;
        }
        int[] index = new int[256];
        for (int i = 0; i < magazine.length(); i++) {
            index[magazine.charAt(i)]++;
        }
        for (int i = 0; i < ransomNote.length(); i++) {
            if (index[ransomNote.charAt(i)]-- <= 0) {
                return false;
            }
        }
        return true;
    }
}
